projects/driver_controls/src/pedal_calibration.c

```diff
@@ -0,0 +1,192 @@
+StatusCode pedal_calibration_init(PedalCalibrationStorage *storage,
+                                  PedalCalibrationSettings *settings) {
+  // TODO(ELEC-382): validate inputs
+  memset(storage, 0, sizeof(*storage));
+  storage->settings = *settings;
+
+  return STATUS_CODE_OK;
+}
+
+// Given the state (full throttle/brake), it finds the min and max raw readings of each channel.
+// These points would become the vertices of the band we want to obtain around the data.
+StatusCode pedal_calibration_process_state(PedalCalibrationStorage *storage,
+                                           PedalCalibrationState state) {
+  if (storage == NULL || state >= NUM_PEDAL_CALIBRATION_STATES) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  bool disabled = critical_section_start();
+  storage->state = state;
+  storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_A] = 0;
+  storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_B] = 0;
+
+  int16_t reading;
+  for (PedalCalibrationChannel channel = PEDAL_CALIBRATION_CHANNEL_A;
+       channel < NUM_PEDAL_CALIBRATION_CHANNELS; channel++) {
+    storage->band[channel][state].min = INT16_MAX;
+    storage->band[channel][state].max = INT16_MIN;
+  }
+  critical_section_end(disabled);
+
+  // Set up callback on channels which would read new values after each conversion.
+  // Since we disable the channel reads as we obtain enough samples, they need to be reconfigured
+  // for the new stage.
+  status_ok_or_return(ads1015_configure_channel(
+      storage->settings.ads1015_storage, storage->settings.adc_channel[PEDAL_CALIBRATION_CHANNEL_A],
+      true, prv_adc_callback, storage));
+  status_ok_or_return(ads1015_configure_channel(
+      storage->settings.ads1015_storage, storage->settings.adc_channel[PEDAL_CALIBRATION_CHANNEL_B],
+      true, prv_adc_callback, storage));
+
+  // Wait until both channels finish sampling.
+  while (storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_A] < PEDAL_CALIBRATION_NUM_SAMPLES &&
+         storage->sample_counter[PEDAL_CALIBRATION_CHANNEL_B] < PEDAL_CALIBRATION_NUM_SAMPLES) {
+  }
+
+  return STATUS_CODE_OK;
+}
+
+static void prv_calc_midline(PedalCalibrationStorage *storage,
+                             PedalCalibrationChannel calib_channel,
+                             ThrottleChannel throttle_channel) {
+  PedalCalibrationRange *range = storage->band[calib_channel];
+  storage->settings.throttle_calibration_data->line[throttle_channel] = (ThrottleLine){
+    .full_brake_reading = (range[PEDAL_CALIBRATION_STATE_FULL_BRAKE].max +
+                           range[PEDAL_CALIBRATION_STATE_FULL_BRAKE].min) /
+                          2,
+    .full_throttle_reading = (range[PEDAL_CALIBRATION_STATE_FULL_THROTTLE].max +
+                              range[PEDAL_CALIBRATION_STATE_FULL_THROTTLE].min) /
+                             2,
+  };
+}
+
+static int16_t prv_calc_range(PedalCalibrationRange *range) {
+  return range[PEDAL_CALIBRATION_STATE_FULL_THROTTLE].max -
```

Comment:
Should these both be max?

My understanding was that the pedal range was:
```
0 ------------------------ 50 ----------------------- 100
Max Brake -- Min Brake -- Coast -- Min Accel -- Max Accel
```

Feel free to correct me if I am wrong.

Reply:
No, this is in the context of voltage. The min/max is not referring to the range of the zone, but the peak-to-peak ripple/measurement error at a specific point. In this case, it's the two endpoints.

You would be correct if this was percentage.